web_fragments/tests/test_fragment.py

```diff
@@ -0,0 +1,264 @@
+        fragment_dict = fragment_dict if fragment_dict else fragment.to_dict()
+        self.assertEqual(fragment_dict['content'], TEST_HTML)
+        self.assertEqual(fragment_dict['js_init_fn'], TEST_JS_INIT_FN)
+        self.assertEqual(fragment_dict['js_init_version'], EXPECTED_JS_INIT_VERSION)
+        self.assertEqual(fragment_dict['json_init_args'], TEST_JSON_INIT_ARGS)
+        self.assertEqual(fragment_dict['resources'], self.test_resources_dict())
+
+    def test_resources_dict(self):
+        """
+        Returns the expected test resources.
+        """
+        return [
+            {
+                'kind': 'text',
+                'data': TEST_CSS,
+                'mimetype': 'text/css',
+                'placement': 'head',
+            },
+            {
+                'kind': 'url',
+                'data': TEST_CSS_URL,
+                'mimetype': 'text/css',
+                'placement': 'head',
+            },
+            {
+                'kind': 'text',
+                'data': TEST_JS,
+                'mimetype': 'application/javascript',
+                'placement': 'foot',
+            },
+            {
+                'kind': 'url',
+                'data': TEST_JS_URL,
+                'mimetype': 'application/javascript',
+                'placement': 'foot',
+            },
+        ]
+
+    def test_to_dict(self):
+        """
+        Tests the to_dict method.
+        """
+        fragment = self.create_test_fragment()
+        fragment_dict = fragment.to_dict()
+        self.validate_fragment(fragment_dict=fragment_dict)
+
+    def test_from_dict(self):
+        """
+        Tests the from_dict method.
+        """
+        test_dict = {
+            'content': TEST_HTML,
+            'resources': self.test_resources_dict(),
+            'js_init_fn': TEST_JS_INIT_FN,
+            'js_init_version': EXPECTED_JS_INIT_VERSION,
+            'json_init_args': TEST_JSON_INIT_ARGS,
+        }
+        fragment = Fragment.from_dict(test_dict)
+        self.validate_fragment(fragment)
+
+    def test_body_html(self):
+        """
+        Tests the body_html method.
+        """
+        fragment = self.create_test_fragment()
+        html = fragment.body_html()
+        self.assertEqual(html, TEST_HTML)
+
+    def test_head_html(self):
+        """
+        Tests the head_html method.
+        """
+        fragment = self.create_test_fragment()
+        html = fragment.head_html()
+        self.assertEqual(
+            html.replace('\n', ''),
+            CSS_EXPECTED_HTML.format(css=TEST_CSS) + CSS_LINK_EXPECTED_HTML.format(css_url=TEST_CSS_URL)
+        )
+
+    def test_foot_html(self):
+        """
+        Tests the foot_html method.
+        """
+        fragment = self.create_test_fragment()
+        html = fragment.foot_html()
+        self.assertEqual(
+            html.replace('\n', ''),
+            JS_EXPECTED_HTML.format(js=TEST_JS) + JS_LINK_EXPECTED_HTML.format(js_url=TEST_JS_URL)
+        )
```

Comment:
1. The logic in the `assertEqual` is entirely non-obvious on a quick (or even slow) read. If you want to keep it as-is, you might consider breaking some of this out into a function indicating its own function with a name indicating what it does. 
2. There is a lot of repeated code between `test_body_html` and here. You might consider making one `test_html` function, and seeing if there's a way to combine the three somehow. 
3. An exact comparison is, again, brittle. If we change the code which handles this (e.g. newline replacement), we'll have many, cascade, cut-and-paste test failures. 

Perhaps something like:

     fragment = self.create_test_fragment()
     foot = fragment.foot_html()
     head = fragment.head_html()
     body = fragment.body_html()
     self._validate_fragment_html(foot, EXPECTED_FOOT_HTML)

`_validate_fragment` would compare for containment, or ignoring new lines, or similar. 

`CSS_EXPECTED_HTML.format(css=TEST_CSS)` is also more code than needed. We never use this any other way. We' be better off just including the css in the global than the cut-and-paste repetition.